core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -335,6 +340,18 @@ void mergeWith(
     // Has another set merged with this?
     assert equivalentSet == null;
 
+    // calls propagateCostImprovements() for RelSubset instances,
+    // whose best should be changed to checks whether that
+    // subset's parents have gotten cheaper.
+    final Set<RelSubset> activeSet = new HashSet<>();
+    for (Map.Entry<RelSubset, RelNode> subsetBestPair : changedSubsets.entrySet()) {
+      RelSubset relSubset = subsetBestPair.getKey();
+      relSubset.propagateCostImprovements(
+          planner, mq, subsetBestPair.getValue(),
+          activeSet);
```

Comment:
Okey, this is definitely needed.

Reply:
Agree, this is one of the key changes in this PR.

Reply:
What's the reason of recomputing otherSubset.best? Do you think its input can be changed at this point? Can you please give a scenario? Thanks.

Reply:
It is required for the case when this `RelSet` contains `RelSubset` which matches to the trait sets of `otherSubset`, so best may be changed (`otherSubset.bestCost.isLt(subset.bestCost)` check) and therefore parent rel nodes should be updated.

Reply:
Thanks. Now I see why. It looks to me that if we move these codes to be after line 363 (renaming parentRel), then we won't need line 373 to line 378. What do you think?